File: go/028 Implement strStr/ImplementstrStr.go

// Implement strStr(). Returns the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.
func strStr(haystack string, needle string) int {
    n, m := len(haystack), len(needle)
    if m > n {return -1} else if m == 0 {return 0}
    for i := 0; i <= n - m; i++ {
        tag := true
        for a, b := 0, i; a < m;  {
            if haystack[b] != needle[a] {
                tag = false
                break
            }
            a += 1
            b += 1
        }
        if tag {return i}
    }
    return -1
}

File: js/028 Implement strStr/ImplementstrStr.js

//Implement strStr(). Returns the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.
/**
 * @param {string} haystack
 * @param {string} needle
 * @return {number}
 */
var strStr = function(haystack, needle) {
    const n = haystack.length, m = needle.length;
    if (m > n) return -1;
    if (m == 0) return 0;
    for (let i = 0; i <= n - m; i++) {
        let tag = true;
        for (let a = 0, b = i; a < m; ) {
            if (haystack[b++] != needle[a++]) {
                tag = false;
                break;
            }
        }
        if (tag) return i;
    }
    return -1;
};
